### ISSUE TITLE
step_by + skip produce incorrect results

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust#55985]
The following snippet demonstrates the problem:
```rust
println!("first");
for i in (0..640).step_by(128) {
    println!("{}", i)
}
println!("second");
for i in (0..640).step_by(128).skip(1) {
    println!("{}", i)
}
println!("third");
for i in (0..640).step_by(128).map(|i| i).skip(1) {
    println!("{}", i)
}
```

It prints:
```
first
0
128
256
384
512
second
128
129
257
385
513
third
128
256
384
512
```

I expect second iterator to produce the same result as for the third one.

**Comment by jethrogb** [comment]
Regression from 1.28.0 to 1.29.0. Introduced in https://github.com/rust-lang/rust/pull/51601

**Comment by m1el** [comment]
The problem is interaction between `StepBy::nth` and `StepBySpecIterator::spec_next`. `spec_next` ignores the inner working of StepBy.

Test-case reduced to using `StepBy` without `Skip`.

```rust
fn main() {
    let mut it = (0_usize..640).step_by(128);
    it.nth(0);
    assert_eq!(it.next(), Some(128));
}
```

**Comment by Emerentius** [comment]
From a cursor look, it seems to me that the problem is that `nth` and `next` are out of sync. StepBy specifically allows [two different kind of semantics](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.step_by) for the underlying iterator, either `next(), nth(step-1), nth(step-1), …` or `advance_n_and_return_first`. If both `next` and `nth` follow the same one, the bug shouldn't occur.

In other words, if `next` is specialized and `nth` overrides the default, then `nth` needs to be specialized as well.
